src/Swarm/Game/Step.hs

```diff
@@ -1617,3 +1623,43 @@ safeExp :: Has (Throw Exn) sig m => Integer -> Integer -> m Integer
 safeExp a b
   | b < 0 = throwError $ CmdFailed Exp "Negative exponent"
   | otherwise = return $ a ^ b
+
+-- | Update the global list of discovered entities, and check for new recipes.
+updateDiscoveredEntities :: (Has (State GameState) sig m, Has (State Robot) sig m) => Entity -> m ()
+updateDiscoveredEntities e = do
+  allDiscovered <- use allDiscoveredEntities
+  if E.contains0plus e allDiscovered
+    then pure ()
+    else do
+      let newAllDiscovered = E.insertCount 9001 e allDiscovered
+      allDevices <- use allInstalledDevices
+      updateAvailableRecipes (newAllDiscovered, allDevices) e
+      allDiscoveredEntities .= newAllDiscovered
+
+-- | Update the global list of installed devices, and check for new recipes.
+updateInstalledDevice :: (Has (State GameState) sig m, Has (State Robot) sig m) => Entity -> m ()
+updateInstalledDevice e = do
+  allDevices <- use allInstalledDevices
+  if E.contains0plus e allDevices
+    then pure ()
+    else do
+      let newAllDevices = E.insertCount 9001 e allDevices
+      -- Check every known entities to see if the new device enables new recipes
+      allDiscovered <- use allDiscoveredEntities
+      traverse_ (updateAvailableRecipes (allDiscovered, newAllDevices) . snd) (elems allDiscovered)
+      allInstalledDevices .= newAllDevices
+
+updateAvailableRecipes :: Has (State GameState) sig m => (Inventory, Inventory) -> Entity -> m ()
+updateAvailableRecipes invs e = do
+  allInRecipes <- use recipesIn
+  let entityRecipes = recipesFor allInRecipes e
+  let canMake recipe = case missingIngredientsFor invs recipe of
+        [] -> True
+        _ -> False
+  let usableRecipes = filter canMake entityRecipes
+  (knownLength, knownRecipes) <- use availableRecipes
+  let newRecipes = filter (`notElem` knownRecipes) usableRecipes
+      newLength = case newRecipes of
+        [] -> knownLength
+        _ -> length newRecipes
+  availableRecipes .= (newLength, newRecipes <> knownRecipes)
```

Comment:
I see two potential sources of inefficiency here:
- Every time we discover a new entity, we iterate through the entire list of recipes to see which ones we can make.
    - I am really not too concerned about this. Trying to do something more clever seems like it would definitely be a case of premature optimization.  One doesn't discover new entities all that often, and I can't imagine the number of recipes will ever grow more than a few thousand.
- For each usable recipe, we do a linear search through the list of known recipes to see if we already know it.
    - This is a little more troubling, since it's quadratic in the number of recipes.  But again, it probably doesn't really make that much difference until we get up to thousands of recipes.

Reply:
Oh yes, that is worrisome, that's the reason why I added two new Inventories to avoid doing this update unless necessary. I'll copy this comment for future reference, when it will become a bottleneck.